filesystem/flysystem/FlysystemAssetStore.php

```diff
@@ -0,0 +1,368 @@
+				throw new \InvalidArgumentException("File already exists at path {$fileID}");
+			}
+
+			// Rename
+			case AssetStore::CONFLICT_RENAME: {
+				foreach($this->fileGeneratorFor($fileID) as $candidate) {
+					// @todo better infinite loop breaking
+					if(!$this->getFilesystem()->has($candidate)) {
+						return $candidate;
+					}
+				}
+
+				throw new \InvalidArgumentException("File could not be renamed with path {$fileID}");
+			}
+
+			// Default to use existing file
+			case AssetStore::CONFLICT_USE_EXISTING:
+			default: {
+				return false;
+			}
+		}
+	}
+
+	/**
+	 * Get an asset renamer for the given filename.
+	 *
+	 * @param string $fileID Adaptor specific identifier for this file/version
+	 * @return AssetNameGenerator
+	 */
+	protected function fileGeneratorFor($fileID){
+		return Injector::inst()->createWithArgs('AssetNameGenerator', array($fileID));
+	}
+
+	/**
+	 * Performs filename cleanup before sending it back.
+	 *
+	 * This name should not contain hash or variants.
+	 *
+	 * @param string $filename
+	 * @return string
+	 */
+	protected function cleanFilename($filename) {
+		// Since we use double underscore to delimit variants, eradicate them from filename
+		return preg_replace('/_{2,}/', '_', $filename);
+	}
+
+	/**
+	 * Given a FileID, map this back to the original filename, trimming variant
+	 *
+	 * @param string $fileID Adaptor specific identifier for this file/version
+	 * @param string $variant Out parameter for any found variant
+	 * @return string
+	 */
+	protected function getOriginalFilename($fileID, &$variant = '') {
+		// Remove variant
+		$original = $fileID;
+		$variant = '';
+		if(preg_match('/^(?<before>((?<!__).)+)__(?<variant>[^\\.]+)(?<after>.*)$/', $fileID, $matches)) {
+			$original = $matches['before'].$matches['after'];
+			$variant = $matches['variant'];
+		}
+
+		// Remove hash
+		return preg_replace(
+			'/(?<hash>[a-zA-Z0-9]{10}\\/)(?<name>[^\\/]+)$/',
+			'$2',
+			$original
+		);
+	}
+
+	/**
+	 * Map file tuple (hash, name, variant) to a filename to be used by flysystem
+	 *
+	 * The resulting file will look something like my/directory/EA775CB4D4/filename__variant.jpg
+	 *
+	 * @param string $hash
+	 * @param string $filename Name of file
+	 * @param string $variant (if given)
+	 * @return string Adaptor specific identifier for this file/version
+	 */
+	protected function getFileID($hash, $filename, $variant = null) {
+		// Since we use double underscore to delimit variants, eradicate them from filename
+		$filename = $this->cleanFilename($filename);
+		$name = basename($filename);
+
+		// Split extension
+		$extension = null;
+		if(($pos = strpos($name, '.')) !== false) {
+			$extension = substr($name, $pos);
+			$name = substr($name, 0, $pos);
+		}
+
+		// Unless in legacy mode, inject hash just prior to the filename
+		if(Config::inst()->get(__CLASS__, 'legacy_filenames')) {
+			$fileID = $name;
+		} else {
+			$fileID = substr($hash, 0, 10) . '/' . $name;
```

Comment:
I think the improvement in the url length is probably justified in this case. Long urls with arbitrary characters obscure the value of the filename.